File: AoC2024/11/main.hs
import GHC.Conc (par, pseq)

part1 :: [Int] -> Int
part1 xs = blinkXDFS xs 25

parseInput :: String -> [Int]
parseInput s = map read (words s)

part2 :: [Int] -> Int
part2 xs = blinkXDFS xs 75

countDigitsHelper :: Int -> Int -> Int
countDigitsHelper 0 y = y
countDigitsHelper x y = countDigitsHelper (x `div` 10) (y +1)

countDigits :: Int -> Int
countDigits x = countDigitsHelper x 0

blinkDFS :: Int -> Int -> Int
blinkDFS _ 0 = 1
blinkDFS 0 depth = blinkDFS 1 (depth - 1)
blinkDFS x depth
    | let digit_count = countDigits x,
      even digit_count =
      let mask = 10 ^ (digit_count `div` 2)
          left = x `div` mask
          right = x `mod` mask
      in blinkDFS left (depth - 1) + blinkDFS right (depth - 1)
    | otherwise = blinkDFS (x * 2024) (depth - 1)

blinkXDFS :: [Int] -> Int -> Int
blinkXDFS xs y = foldr (\ x -> (+) (blinkDFS x y)) 0 xs

-- parallel blink for each number
pblinkXDFS :: [Int] -> Int -> Int
pblinkXDFS [] _ = 0
pblinkXDFS (x:xs) y = par n1 (pseq n2 (n1 + n2))
    where n1 = blinkDFS x y
          n2 = pblinkXDFS xs y

main :: IO()
main = do
    input <- readFile "input"
    let parsed = parseInput input
    putStrLn ("Part 1: " ++ show (part1 parsed))
    putStrLn ("Part 2: " ++ show (part2 parsed))

File: AoC2024/11/main.py
from functools import lru_cache

inp = []
with open("input") as f:
    inp = [int(num) for num in f.read().split(" ")]


def count_digits(x):
    count = 0
    while x > 0:
        x //= 10
        count += 1
    return count


@lru_cache(maxsize=None)
def blinkXTimes(stone, depth=25):
    while depth > 0:
        if stone == 0:
            stone = 1
            depth -= 1
            continue
        d_count = count_digits(stone)
        if d_count % 2 == 0:
            mask = 10 ** (d_count // 2)
            l = stone // mask
            r = stone % mask
            return blinkXTimes(l, depth - 1) + blinkXTimes(r, depth - 1)

        stone *= 2024
        depth -= 1
    return 1


print("Part 1: ", sum([blinkXTimes(x) for x in inp]))
print("Part 2: ", sum([blinkXTimes(x, 75) for x in inp]))
